README.md
Minimal gRPC Server
===================
This repo provides libraries that run a Google gRPC server synchronously, which will exit upon receiving SIGTERM, SIGQUIT, or SIGINT.


CHANGELOG.md
# Changelog

## [0.7.2](https://github.com/kurtosis-tech/minimal-grpc-server/compare/0.7.1...0.7.2) (2023-07-10)


### Features

* added grpcweb proxy that allows me to access grpc endpoints from the ui ([#39](https://github.com/kurtosis-tech/minimal-grpc-server/issues/39)) ([90b674a](https://github.com/kurtosis-tech/minimal-grpc-server/commit/90b674acb26995ddff5665918059aeff8db71797))

## [0.7.1](https://github.com/kurtosis-tech/minimal-grpc-server/compare/0.7.0...0.7.1) (2023-03-17)


### Features

* Add possibility to enable HTTPS on GRPC server ([#35](https://github.com/kurtosis-tech/minimal-grpc-server/issues/35)) ([4b160f0](https://github.com/kurtosis-tech/minimal-grpc-server/commit/4b160f0d17db8a7d952e9a5fc6c26251ea176cc2))

## [0.7.0](https://github.com/kurtosis-tech/minimal-grpc-server/compare/0.6.2...0.7.0) (2023-03-16)


### ⚠ BREAKING CHANGES

* RunUntilStopped now accepts a chan <-struct{} instead of chan <-interface{} ([#33](https://github.com/kurtosis-tech/minimal-grpc-server/issues/33))

### Code Refactoring

* RunUntilStopped now accepts a chan &lt;-struct{} instead of chan <-interface{} ([#33](https://github.com/kurtosis-tech/minimal-grpc-server/issues/33)) ([1f3f722](https://github.com/kurtosis-tech/minimal-grpc-server/commit/1f3f72269eacff45582f24cef7524c582011d96a))

## [0.6.2](https://github.com/kurtosis-tech/minimal-grpc-server/compare/0.6.1...0.6.2) (2023-03-16)


### Bug Fixes

* bump ansi-regex from 3.0.0 to 3.0.1 in /typescript ([#23](https://github.com/kurtosis-tech/minimal-grpc-server/issues/23)) ([4adf4f0](https://github.com/kurtosis-tech/minimal-grpc-server/commit/4adf4f054260ba0c8aa1cefb1b86acab4e13d1ab))
* bump json5 from 2.2.0 to 2.2.3 in /typescript ([#28](https://github.com/kurtosis-tech/minimal-grpc-server/issues/28)) ([7b3b453](https://github.com/kurtosis-tech/minimal-grpc-server/commit/7b3b4534e987ddc55879a15e380d1b7f1aa8ee38))
* bump minimist from 1.2.5 to 1.2.7 in /typescript ([#25](https://github.com/kurtosis-tech/minimal-grpc-server/issues/25)) ([c16a423](https://github.com/kurtosis-tech/minimal-grpc-server/commit/c16a423b5d1ca551acee20fb03a3a5d2730cec25))

## 0.6.1

* Migrate repo to use internal tool `kudet` for release process
* Migrate `develop` into `master`
* 
## 0.6.0
### Changes
* Switched the Typescript library to use `@grpc/grpc-js` rather than the `grpc` package, as the `grpc` package is deprecated

### Breaking Changes
* The service interfaces that are passed in to the server must be of type `@grpc/grpc-js` rather than `grpc`

## 0.5.0
### Removals
* Removed the "protocol" arg to the Go library, because gRPC can only run on TCP

### Breaking Changes
* `NewMinimalGRPCServer` constructors in the Go library no longer take in a protocol

## 0.4.0
### Features
* Added a `runUntilStopped` method to the server, allowing the server to be stopped using an arbitrary event (rather than only interrupts)
* Added `runUntilStopped` tests to Golang & Typescript

### Changes
* Renamed `run` to `runUntilInterrupted`

### Fixes
* Fixed a bug where even if the Typescript server shut down correctly, it would still wait for the hard stop timeout
* Fixed a bug in the TS lib where the server would go through the hard stop flow if the server actually stopped correctly, and vice versa

### Breaking Changes
* Renamed `run` to `runUntilInterrupted`
* The Golang library's listen port is now a `uint16`
* The Typescript library now requires Node >= `16.13.0`

## 0.3.8
### Fixes
* `stacktrace.Propagate` panics when receiving a `nil` cause

## 0.3.7
### Features
* Added logging to all requests made to users of the Go version of this library
    * **NOTE:** This is NOT enabled for Typescript because server-side interceptors aren't supported unfortunately: https://github.com/grpc/grpc-node/issues/419

### Fixes
* Added `go build ./...` to Go buildscript, as some compile errors weren't getting caught

## 0.3.6
### Fixes
* Fixed a bug where the gRPC server was binding on the incorrect IP

## 0.3.5
### Fixes
* Fixed a bug where the bind required `127.0.0.1` specified

## 0.3.4
### Fixes
* Fixed a bug in the way we created the insecure server credentials

## 0.3.3
### Fixes
* Unpin Node engine, using `>=14.17.0` now

## 0.3.2
### Features
* Enabled Typescript strict mode for safer code

### Fixes
* Fixed bugs that popped out when strict mode was enabled

## 0.3.1
### Fixes
* Export `TypedServerOverride`

## 0.3.0
### Fixes
* Added a workaround for gRPC's stupid "unimplemented server" requirement, which messes everything up in Typescript

### Breaking Changes
* The Typescript `MinimalGRPCServer` now takes in service registration functions that accept `TypedServerOverride` rather than `grpc.Server`
    * Users should make their server implementation class implement `KnownKeysOnly<ITheUserServiceServer>` rather than `ITheUserServiceServer`, and in their registration functions call `typedServerOverride.addTypedService` instead of `.addService`

## 0.2.2
### Changes
* Switch to using productized docs-checker orb

### Features
* Added a Typescript version

## 0.2.1
### Features
* Added CircleCI checks

## 0.2.0
### Changes
* The Go module is now one layer deeper

### Breaking Changes
* The Go module is now one layer deeper, at `github.com/kurtosis-tech/minimal-grpc-server/golang`
    * Users should append the `/golang` to the end of all their module imports

## 0.1.0
* Initial release


golang/server/server_test.go
package server

import (
	"github.com/stretchr/testify/assert"
	"google.golang.org/grpc"
	"testing"
	"time"
)

const (
	listenPort      = uint16(9003)
	stopGracePeriod = 10 * time.Second
)

func TestManualShutdown(t *testing.T) {
	server := NewMinimalGRPCServer(listenPort, stopGracePeriod, []func(*grpc.Server){})

	stopChan := make(chan struct{})
	runResultChan := make(chan error)
	go func() {
		runResultChan <- server.RunUntilStopped(stopChan)
	}()
	select {
	case err := <-runResultChan:
		t.Fatalf("The server unexpectedly returned a result before we sent a signal to stop it:\n%v", err)
	case <-time.After(1 * time.Second):
		// Nothing
	}
	stopChan <- struct{}{}

	maxWaitForServerToStop := 1 * time.Second
	var errAfterStop error
	select {
	case errAfterStop = <-runResultChan:
		// Assignment is all we have to do
	case <-time.After(maxWaitForServerToStop):
		t.Fatalf("Expected the server to have stopped after %v, but it didn't", maxWaitForServerToStop)
	}
	assert.NoError(t, errAfterStop, "Expected the server to exit without an error but it threw the following:\n%v", errAfterStop)
}


typescript/test/server_test.ts
import { strict as assert } from "assert";
import { Result } from 'neverthrow';
import { MinimalGRPCServer, TypedServerOverride } from "../src/server";
import { setTimeout } from "timers/promises";

const LISTEN_PORT: number = 9003;
const STOP_GRACE_PERIOD_SECONDS: number = 10;

describe(MinimalGRPCServer.name, function() {
    describe('#runUntilStopped', function() {
        it('should start the server, let it run for a bit, then stop it without issue', async function() {
            const serviceRegistrationFuncs: { (server: TypedServerOverride): void; }[] = [];
            const server = new MinimalGRPCServer(LISTEN_PORT, STOP_GRACE_PERIOD_SECONDS, serviceRegistrationFuncs);
            var stopFunction;
            const stopPromise: Promise<null> = new Promise((resolve) => { stopFunction = resolve; })

            const serverResultPromise: Promise<Result<null, Error>> = server.runUntilStopped(stopPromise);

            const firstTimeoutPromise: Promise<null> = setTimeout(1000, undefined);
            const firstRaceResult: Result<null, Error> | undefined = await Promise.race([serverResultPromise, firstTimeoutPromise]);
            assert.equal(firstRaceResult, undefined, `The server unexpectedly returned a result before we sent a signal to stop it:\n${firstRaceResult}`);
            
            stopFunction();

            const maxWaitForServerToStopMillis: number = 1000;
            const secondTimeoutPromise: Promise<null> = setTimeout(maxWaitForServerToStopMillis, null);
            const secondRaceResult: Result<null, Error> | undefined = await Promise.race([serverResultPromise, secondTimeoutPromise]);
            if (secondRaceResult === undefined) {
                assert.fail(`Expected the server to have stopped after ${maxWaitForServerToStopMillis}ms, but it didn't`);
            }
            if (secondRaceResult.isErr()) {
                assert.fail(`Expected the server to exit without an error but it threw the following:\n${secondRaceResult.error}`);
            }
        })
    })
});

golang/scripts/build.sh
#!/usr/bin/env bash
# 2021-07-08 WATERMARK, DO NOT REMOVE - This script was generated from the Kurtosis Bash script template

set -euo pipefail   # Bash "strict mode"
script_dirpath="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
lang_root_dirpath="$(dirname "$script_dirpath}")"


# ==================================================================================================
#                                             Constants
# ==================================================================================================

# ==================================================================================================
#                                       Arg Parsing & Validation
# ==================================================================================================

# ==================================================================================================
#                                             Main Logic
# ==================================================================================================
cd "${lang_root_dirpath}"
go build ./...
go test ./...


golang/server/server.go
package server

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"github.com/improbable-eng/grpc-web/go/grpcweb"
	"github.com/kurtosis-tech/stacktrace"
	"github.com/sirupsen/logrus"
	"github.com/soheilhy/cmux"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"
)

const (
	// gRPC servers can ONLY run on TCP
	// See https://stackoverflow.com/questions/65111895/using-udp-in-grpc
	listenProtocol = "tcp"
)

type MinimalGRPCServer struct {
	listenPort               uint16
	stopGracePeriod          time.Duration // How long we'll give the server to stop after asking nicely before we kill it
	serviceRegistrationFuncs []func(*grpc.Server)

	// serverCert is optional. If present, server will run HTTPS
	serverCert *tls.Certificate
	// certPool is optional and should be present only if serverCert is present.
	// If present, server will run two-way SSL
	certPool *x509.CertPool
}

// NewMinimalGRPCServer Creates a minimal gRPC server but doesn't start it
// The service registration funcs will be applied, in order, to register services with the underlying gRPC server object
func NewMinimalGRPCServer(listenPort uint16, stopGracePeriod time.Duration, serviceRegistrationFuncs []func(*grpc.Server)) *MinimalGRPCServer {
	return &MinimalGRPCServer{
		listenPort:               listenPort,
		stopGracePeriod:          stopGracePeriod,
		serviceRegistrationFuncs: serviceRegistrationFuncs,
		certPool:                 nil,
		serverCert:               nil,
	}
}

// NewMinimalHttpsGRPCServer is similar to NewMinimalGRPCServer but accepts SSL CA and certificate to enable HTTPS
func NewMinimalHttpsGRPCServer(listenPort uint16, stopGracePeriod time.Duration, certPool *x509.CertPool, serverCert *tls.Certificate, serviceRegistrationFuncs []func(*grpc.Server)) *MinimalGRPCServer {
	return &MinimalGRPCServer{
		listenPort:               listenPort,
		stopGracePeriod:          stopGracePeriod,
		serviceRegistrationFuncs: serviceRegistrationFuncs,
		certPool:                 certPool,
		serverCert:               serverCert,
	}
}

// RunUntilInterrupted runs the server synchronously until an interrupt signal is received
func (server MinimalGRPCServer) RunUntilInterrupted() error {
	// Signals are used to interrupt the server, so we catch them here
	termSignalChan := make(chan os.Signal, 1)
	signal.Notify(termSignalChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
	serverStopChan := make(chan struct{}, 1)
	go func() {
		<-termSignalChan
		interruptSignal := struct{}{}
		serverStopChan <- interruptSignal
	}()
	if err := server.RunUntilStopped(serverStopChan); err != nil {
		return stacktrace.Propagate(err, "An error occurred running the server using the interrupt channel for stopping")
	}
	return nil
}

// RunUntilStopped runs the server synchronously until a signal is received on the given channel
func (server MinimalGRPCServer) RunUntilStopped(stopper <-chan struct{}) error {
	loggingInterceptorFunc := func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
		grpcMethod := info.FullMethod
		logrus.Debugf("Received gRPC request to method '%v' with args:\n%+v", grpcMethod, req)
		resp, err := handler(ctx, req)
		if err != nil {
			logrus.Debugf("gRPC request to method '%v' failed with error:\n%v", grpcMethod, err)
		} else {
			logrus.Debugf("gRPC request to method '%v' succeeded with response:\n%+v", grpcMethod, resp)
		}
		return resp, err
	}
	serverOptions := []grpc.ServerOption{
		grpc.UnaryInterceptor(loggingInterceptorFunc),
	}

	// Add TLS credentials to server options if certificate have been provided
	tlsCredentialsMaybe := server.loadTlsCredentials()
	if tlsCredentialsMaybe != nil {
		serverOptions = append(serverOptions, grpc.Creds(tlsCredentialsMaybe))
	}
	grpcServer := grpc.NewServer(serverOptions...)
	for _, registrationFunc := range server.serviceRegistrationFuncs {
		registrationFunc(grpcServer)
	}

	listenAddressStr := fmt.Sprintf("0.0.0.0:%v", server.listenPort)
	listener, err := net.Listen(listenProtocol, listenAddressStr)
	if err != nil {
		return stacktrace.Propagate(
			err,
			"An error occurred creating the listener on %v/%v",
			listenProtocol,
			server.listenPort,
		)
	}

	maybeErrorResultChan := make(chan error)
	mux := cmux.New(listener)
	grpcWebListener := mux.Match(cmux.HTTP1Fast())
	grpcListener := mux.Match(cmux.Any())

	// grpcweb is a proxy which enables ui to make requests to grpc server using kurtosis-sdk
	// it converts http/1 request to http/2 and vice-versa under the hood
	grpcWebServer := grpcweb.WrapServer(grpcServer, grpcweb.WithOriginFunc(func(origin string) bool { return true }))

	go func() {
		if resultErr := grpcServer.Serve(grpcListener); resultErr != nil {
			logrus.Debugf("error ocurred while creating grpc server: %v", resultErr)
			maybeErrorResultChan <- resultErr
		}
	}()

	go func() {
		httpServer := &http.Server{
			Handler: http.Handler(grpcWebServer),
		}
		if resultErr := httpServer.Serve(grpcWebListener); resultErr != nil {
			logrus.Debugf("error ocurred while creating grpcweb server: %v", resultErr)
			maybeErrorResultChan <- resultErr
		}
	}()

	go func() {
		if resultErr := mux.Serve(); resultErr != nil {
			logrus.Debugf("error ocurred while creating mux server: %v", resultErr)
			maybeErrorResultChan <- resultErr
		}
	}()

	// Wait until we get a shutdown signal
	<-stopper

	serverStoppedChan := make(chan interface{})
	go func() {
		grpcServer.Stop()
		mux.Close()
		serverStoppedChan <- nil
	}()
	select {
	case <-serverStoppedChan:
		logrus.Debug("gRPC server has exited gracefully")
	case <-time.After(server.stopGracePeriod):
		logrus.Warnf("gRPC server failed to stop gracefully after %v; hard-stopping now...", server.stopGracePeriod)
		grpcServer.Stop()
		mux.Close()
		logrus.Debug("gRPC server was forcefully stopped")
	}

	if err := <-maybeErrorResultChan; err != nil {
		// Technically this doesn't need to be an error, but we make it so to fail loudly
		// this is expected behaviour observed via cmux
		gracefulExit := isGracefulExit(err)
		if !gracefulExit {
			return stacktrace.Propagate(err, "gRPC server returned an error after it was done serving")
		}
	}

	return nil
}
func (server MinimalGRPCServer) loadTlsCredentials() credentials.TransportCredentials {
	if server.serverCert == nil {
		// No certificate provided, will use HTTP
		return nil
	}
	if server.certPool == nil {
		// with no cert pool, 1 way SSL will be enabled, no need for CA
		tlsConfig := &tls.Config{
			Certificates: []tls.Certificate{
				*server.serverCert,
			},
			ClientAuth: tls.NoClientCert,
		}
		return credentials.NewTLS(tlsConfig)
	}
	// 2 ways SSL enabled - Load CA and set ClientAuth to "Require And Verify"
	tlsConfig := &tls.Config{
		Certificates: []tls.Certificate{
			*server.serverCert,
		},
		ClientAuth: tls.RequireAndVerifyClientCert,
		ClientCAs:  server.certPool,
	}
	return credentials.NewTLS(tlsConfig)
}

// look at the comment below for more info
//https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/server.go
func isGracefulExit(err error) bool {
	switch err {
	case nil, http.ErrServerClosed, cmux.ErrListenerClosed, cmux.ErrServerClosed:
		// do nothing, normal exit
		return true
	default:
		if strings.Contains(err.Error(), "use of closed network connection") {
			return true
		}
		return false
	}
}


scripts/update-package-versions.sh
#!/usr/bin/env bash
# 2021-07-08 WATERMARK, DO NOT REMOVE - This script was generated from the Kurtosis Bash script template

set -euo pipefail   # Bash "strict mode"
script_dirpath="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
root_dirpath="$(dirname "${script_dirpath}")"

# ==================================================================================================
#                                             Constants
# ==================================================================================================
SUPPORTED_LANGS_FILENAME="supported-languages.txt"
UPDATE_VERSION_IN_FILE_SCRIPT_FILENAME="update-version-in-file.sh"
SKIP_KEYWORD="SKIP" # Certain languages don't need any package replacement; they will use this keyword for their package file and replacement pattern

# Per-language package filepaths that need version, RELATIVE TO THE LANG ROOT, that need version replacement
declare -A PACKAGE_FILE_RELATIVE_FILEPATHS

declare -A PACKAGE_VERSION_PRINTF_PATTERNS

# Golang -- empty because Go doesn't need any package version replacing
PACKAGE_FILE_RELATIVE_FILEPATHS["golang"]="${SKIP_KEYWORD}"
PACKAGE_VERSION_PRINTF_PATTERNS["golang"]="${SKIP_KEYWORD}"

# Typescript
PACKAGE_FILE_RELATIVE_FILEPATHS["typescript"]="package.json"
PACKAGE_VERSION_PRINTF_PATTERNS["typescript"]="\"version\": \"%s\""

API_DIRNAME="api"

# ==================================================================================================
#                                       Arg Parsing & Validation
# ==================================================================================================
show_helptext_and_exit() {
    echo "Usage: $(basename "${0}") new_version"
    echo ""
    echo "  new_version         The new version that the package files should contain"
    echo ""
    exit 1  # Exit with an error so that if this is accidentally called by CI, the script will fail
}

new_version="${1:-}"

if [ -z "${new_version}" ]; then
    echo "Error: No new version provided" >&2
    show_helptext_and_exit
fi



# ==================================================================================================
#                                             Main Logic
# ==================================================================================================
project_dirpath="${root_dirpath}"
# If the repository contains the API directory in the rootpath the project path will contains this directory too
api_dirpath="${root_dirpath}/${API_DIRNAME}"
if [ -d "${api_dirpath}" ]; then
  project_dirpath="${root_dirpath}/${API_DIRNAME}"
fi

echo "Updating versions in package files for all supported languages..."
supported_langs_filepath="${project_dirpath}/${SUPPORTED_LANGS_FILENAME}"
if ! [ -f "${supported_langs_filepath}" ]; then
    echo "Error: No supported langs file exists at '${supported_langs_filepath}'" >&2
    exit 1
fi
for lang in $(cat "${supported_langs_filepath}"); do
    to_update_relative_filepath="${PACKAGE_FILE_RELATIVE_FILEPATHS["${lang}"]}"
    if [ -z "${to_update_relative_filepath}" ]; then
        echo "Error: No relative filepath to a package file that needs updating was found for language '${lang}'; this $(basename "${0}") script needs to be updated with this information" >&2
        exit 1
    fi
    if [ "${to_update_relative_filepath}" == "${SKIP_KEYWORD}" ]; then
        continue
    fi

    to_update_abs_filepath="${project_dirpath}/${lang}/${to_update_relative_filepath}"

    replacement_pattern="${PACKAGE_VERSION_PRINTF_PATTERNS["${lang}"]}"
    if [ -z "${replacement_pattern}" ]; then
        echo "Error: No replacement pattern was found for language '${lang}'; this $(basename "${0}") script needs to be updated with this information" >&2
        exit 1
    fi
    if [ "${replacement_pattern}" == "${SKIP_KEYWORD}" ]; then
        continue
    fi

    if ! $(kudet update-version-in-file "${to_update_abs_filepath}" "${replacement_pattern}" "${new_version}"); then
        echo "Error: An error occurred setting new version '${new_version}' in '${lang}' package file '${to_update_abs_filepath}' using pattern '${replacement_pattern}'" >&2
        exit 1
    fi
done
echo "Successfully updated the versions in the package files for all supported languages"

typescript/src/server.ts
import * as grpc from '@grpc/grpc-js';
import * as log from 'loglevel';
import { Result, ok, err } from 'neverthrow';
import { setTimeout } from "timers/promises";

const BIND_IP: string = "0.0.0.0";
const MILLIS_IN_SECOND: number = 1000;
const INTERRUPT_SIGNAL: string = "SIGNINT";
const QUIT_SIGNAL: string = "SIGQUIT";
const TERM_SIGNAL: string = "SIGTERM";

// ====================================== NOTE ========================================================= 
// All this rigamarole is necessary due to gRPC's stupid "unimplemented server" requirements
// See:
// - https://github.com/agreatfool/grpc_tools_node_protoc_ts/issues/79
// - https://github.com/agreatfool/grpc_tools_node_protoc_ts/blob/master/doc/server_impl_signature.md
//
// ====================================== NOTE =========================================================
type KnownKeys<T> = {
    [K in keyof T]: string extends K ? never : number extends K ? never : K
} extends { [_ in keyof T]: infer U } ? U : never;

export type KnownKeysOnly<T extends Record<any, any>> = Pick<T, KnownKeys<T>>;

export class TypedServerOverride extends grpc.Server {
    addTypedService<TypedServiceImplementation extends Record<any,any>>(service: grpc.ServiceDefinition, implementation: TypedServiceImplementation): void {
        this.addService(service, implementation);
    }
}

export class MinimalGRPCServer {
    private readonly listenPort: number;
    private readonly stopGracePeriodSeconds: number;    // How long we'll give the server to stop after asking nicely before we kill it
    private readonly serviceRegistrationFuncs: { (server: TypedServerOverride): void; }[]

    // Creates a minimal gRPC server but doesn't start it
    // The service registration funcs will be applied, in order, to register services with the underlying gRPC server object
    constructor(
        listenPort: number,
        stopGracePeriodSeconds: number,
        // NOTE: Users wanting to register services declared with class style should use server.addTypedService,
        //  and pass in a class that implements KnownKeysOnly<IYourServiceServer>
        serviceRegistrationFuncs: { (server: TypedServerOverride): void; }[]
    ) {
        this.listenPort = listenPort;
        this.stopGracePeriodSeconds = stopGracePeriodSeconds;
        this.serviceRegistrationFuncs = serviceRegistrationFuncs;
    }

    // Runs the server synchronously until an interrupt signal is received
    public async runUntilInterrupted(): Promise<Result<null, Error>> {
        // Signals are used to interrupt the server, so we catch them here
        const signalsToHandle: Array<string> = [INTERRUPT_SIGNAL, QUIT_SIGNAL, TERM_SIGNAL];
        const signalReceivedPromises: Array<Promise<null>> = signalsToHandle.map((signal) => {
            return new Promise((resolve, _unusedReject) => {
                process.on(signal, () => {
                    resolve(null);
                });
            });
        });
        const anySignalReceivedPromise: Promise<null> = Promise.race(signalReceivedPromises);
        const runResult: Result<null, Error> = await this.runUntilStopped(anySignalReceivedPromise);
        if (runResult.isErr()) {
            return err(runResult.error);
        }
        return ok(null);
    }

    // Runs the server synchronously until the given promise is resolved
    public async runUntilStopped(stopper: Promise<null>): Promise<Result<null, Error>> {
        // NOTE: This is where we'd want to add server call interceptors to log the request & response...
        // ...but they're not supported: https://github.com/grpc/grpc-node/issues/419
        // As of 2021-09-20, this is a difference from the Go version!
        const grpcServer: TypedServerOverride = new TypedServerOverride();

        for (let registrationFunc of this.serviceRegistrationFuncs) {
            registrationFunc(grpcServer);
        }

        const listenUrl: string = BIND_IP + ":" + this.listenPort;
        const bindPortPromise: Promise<Result<number, Error>> = new Promise((resolve) => {
            grpcServer.bindAsync(listenUrl, grpc.ServerCredentials.createInsecure(), (error: Error | null, portNumber: number) => {
                if (error === null) {
                    resolve(ok(portNumber));
                } else {
                    resolve(err(error));
                }
            });
        })
        const bindPortResult = await bindPortPromise;
        if (bindPortResult.isErr()) {
            return err(bindPortResult.error);
        }
        grpcServer.start();

        await stopper;

        const tryShutdownPromise: Promise<Result<null, Error>> = new Promise((resolve, _unusedReject) => {
            grpcServer.tryShutdown(() => {
                resolve(ok(null));
            })
        })
        const timeoutAbortController: AbortController = new AbortController();
        const timeoutPromise: Promise<Result<null, Error>> = setTimeout(
            this.stopGracePeriodSeconds * MILLIS_IN_SECOND,
            err(new Error("gRPC server failed to stop gracefully after waiting for " + this.stopGracePeriodSeconds + "s")),
            {
                signal: timeoutAbortController.signal,
            },
        );
        const gracefulShutdownResult: Result<null, Error> = await Promise.race([tryShutdownPromise, timeoutPromise]);
        if (gracefulShutdownResult.isOk()) {
            log.debug("gRPC server has exited gracefully");
        } else {
            log.warn("gRPC server failed to stop gracefully after " + this.stopGracePeriodSeconds + "s; hard-stopping now...");
            grpcServer.forceShutdown();
            log.debug("gRPC server was forcefully stopped");
        }
        // If the timeout is still running (i.e. in the event of a successful shutdown), kill the timeout thread
        //  so that the Node engine doesn't block
        timeoutAbortController.abort();
        return ok(null);
    }
}
